File: leetcode/MaximumProductSubarray/maxProduct.java
public class Solution {
    public int maxProduct(int[] A) {
        int len = A.length;
        if (len == 0) return 0;
        int max[] = new int[len], min[] = new int[len], ans = A[0];
        max[0] = min[0] = A[0];
        for (int i = 1; i < len; i++) {
            max[i] = Math.max(max[i - 1] * A[i], Math.max(A[i], min[i - 1] * A[i]));
            min[i] = Math.min(min[i - 1] * A[i], Math.min(A[i], max[i - 1] * A[i]));
            ans = Math.max(ans, max[i]);
        }
        return ans;
    }
}

File: leetcode/MaximumProductOfWordLenghts/maxProduct.cpp
class Solution {
public:
    int maxProduct(vector<string>& words) {
        int ans = 0;
        sort(words.begin(), words.end(), [](const string &a, const string &b) -> bool {
            return a.size() > b.size();
        });
        
        unordered_set<int> set;
        for (int i = 0; i < int(words.size()) - 1; i++) {
            set.clear();
            for (char c : words[i]) set.insert(c - 'a');
            for (int j = 1; j < words.size(); j++) {
                bool valid = true;
                for (char c : words[j]) {
                    if (set.find(c - 'a') != set.end()) {
                        valid = false;
                        break;
                    }
                }
                if (valid) {
                    int newLen = int(words[i].size() * words[j].size());
                    ans = max(ans, newLen);
                    break;
                }
            }
        }
        return ans;
    }
};
